Disable openresty flags that do not work on older openssl versions

eskerda opened the pull request:
* Disables openresty build flags that do not work with older openssl versions.
* Sends build output to stdout to clear up stderr. Open to a better approach to make the command less verbose.

dndx commented on kong-ngx-build:
```diff
@@ -272,17 +272,20 @@ main() {
 
         OPENRESTY_OPTS+=(${NGINX_EXTRA_MODULES[@]})
 
+        if [[ $OPENSSL_VER != 1.0.* ]]; then
```
This should be dependent on OpenResty version, not OpenSSL. I think it is `!= 1.11.*`?

eskerda replied:
You are right, my bad. So, according to this http://nginx.org/en/CHANGES they come in 1.11.4 and 1.11.5 respectively.

Should we try to do something clever with the versions or is it too much? The idea could be to only add build flags for versions that support it, and for each flag would come annotated with the version it came in.

Version parsing is always problematic though

thibaultcha replied:
Having them in branches inlined here is good enough and straightforward; will you have some time to update this PR @eskerda?

eskerda replied:
> Having them in branches inlined here is good enough and straightforward; will you have some time to update this PR @eskerda?

I meant, if we should try something that will allow us to ask "VERSION > 1.11.5". I have this `version.sh` partly cleaned up from `kong-upgrade-tests`, but still I am sure will miss something


See: [version.sh.txt](https://github.com/Kong/openresty-build-tools/files/3315550/version.sh.txt)

eskerda replied:
For example, kong `0.13.1` will still fail, since it uses openresty `1.11.2.5`

thibaultcha replied:
@eskerda It'd make sense to have such comparisons in this script, I agree. If we could avoid a new file and embed those utility functions in `kong-ngx-build` (and keep them as concise as possible), that'd be great though.

thibaultcha commented:
@eskerda Will you also rebase this PR on top of the latest master branch? I merged #19 first given the ratio effort for rebasing it compared to this PR. Thank you!

thibaultcha commented on kong-ngx-build:
```diff
@@ -421,32 +426,32 @@ show_usage() {
 err() {
   builtin echo -en "\033[1;31m"
-  echo "ERR: $@"
+  >&2 echo "ERR: $@"
```
Why not also redirecting the color codes to stderr? It doesn't feel like this change accounts for situations in which stdout and stderr are split.

I suggest using `-t` (`man 1 test`) to check if stderr is a tty, and only output those color codes if that is the case. Output redirected to files doesn't need color.

eskerda replied:
I think for the time being I will leave the non verbose output for another PR. Totally missed the color echo, good point.

eskerda commented:
@dndx @thibaultcha I added version checks, conditional openresty flags.

Also added a very draft version of a `.travis.yml` file. At least it tries to validate the command runs, for some fixed versions. Completely open to improvement / moving it to another PR

thibaultcha commented on kong-ngx-build:
```diff
@@ -387,6 +387,48 @@ main() {
 
+parse_version() {
+    # in foo-x.y.z-foobar captures x y z
+    local re='[^0-9]*([0-9a-zA-Z]*)(?:[.]([0-9a-zA-Z]*))?(?:[.]([0-9a-zA-Z]*))?.*'
```
The last `.*` match and quantifier seem unnecessary.

eskerda replied:
👍 the regex can use some rethinking

thibaultcha commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+jobs:
+  include:
+    # Known builds we want to succeed
+    - script: ./kong-ngx-build --prefix prefix --openresty 1.11.2.5 --openssl 1.0.2n --luarocks 2.4.3 --no-openresty-patches
```
To verify that the `--no-openresty-patches` command succeeded, we should ensure that `work/openresty-1.11.2.5/bundle/.patch_applied` does not exist.

eskerda replied:
I was seeing these more like green path integration tests. IIRC the no-openresty-patches flag was needed for builds with openssl versions 1.0.x to succeed.

thibaultcha replied:
The flag was added to build OpenResty without the patches. We don’t apply any patch to OpenSSL ourselves. There are no patches for OpenSSL 1.11.2.5, but I would expect future maintainers of this test harness to duplicate the test jobs and increase the matrix size. Among these jobs will be OpenResty 1.13.6.* builds with and without patches. All builds without patches should perform explicit checks that those were not applied imho, regardless of their version (since the version number comes from the test matrix). Again, not a blocker so we can take care of this later.

eskerda replied:
Still looking at this one

thibaultcha commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+  include:
+    # Known builds we want to succeed
+    - script: ./kong-ngx-build --prefix prefix --openresty 1.11.2.5 --openssl 1.0.2n --luarocks 2.4.3 --no-openresty-patches
+    - script: ./kong-ngx-build --prefix prefix --openresty 1.13.6.2 --openssl 1.1.1c --luarocks 3.1.3
```
I wouldn't be against some test cases for the version comparison utilities... Especially considering the tricky cases such as letters in OpenSSL version schemes and 4 digits versioning schemes for OpenResty. But ultimately not a blocker atm if you don't have the time.

eskerda replied:
If we add a check on `basename $0` before calling the main function, I think ultimately we could source the script and test the version functions separately.

thibaultcha commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+     - curl
+     - libpcre3
+     - libpcre3-dev
+     - zlib1g-dev
```
We should try reduce this list as much as possible. Maybe we can try:
```
- build-essential
- perl
- curl
```

libpcre can also be added, with our jobs ensuring that it picks up the system's lib, if not specified as the `--pcre` argument to `kong-ngx-build`.

eskerda replied:
In a docker starting from xenial I didn't manage to get a successful build until I installed all these packages. Will try again

thibaultcha replied:
We build ngx_lua from source with many less packages, so worth a try imho.

eskerda replied:
True, the base docker image != generic travis, my bad

thibaultcha commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+sudo: required
+
+language: generic
+
```
Let's also add:

```
compiler:
  - gcc
  - clang
```

eskerda commented on kong-ngx-build:
```diff
@@ -289,7 +289,7 @@ main() {
           )
 
-          if [[ $OPENSSL_VER == 1.1.* ]]; then
+          if version_eq $OPENSSL_VER 1.1; then
```
* These `1.1.*` checks are a bit confusing. I thought it was being checked for exactly openssl versions `1.1.x`, and not `1.1.x` and beyond (`1.2`, `2.0`). Is that the case?
* Any comments on the version_eq check ignoring revisions if no revision_b has been provided? I thought it was useful to check things like `1.1.3` is `1.1.*`ish, but if it's not used, I would prefer to revert it to the straightforward equality.
* 👍 on `lte` and `gte`

thibaultcha replied:
* This is just because OpenSSL 1.1.0 renamed the flag. It is most likely the flag will stick to this name going forward. 
* I don’t have any strong feelings. That’s fine imho. Although here we can say `>= 1.1.0`

eskerda commented:
* Improved version parsing. It now generates a list of version components. I didn't fixup the commits yet, since I want to know if this parsing method is any better. It allows for any number of version components
* Added _lte and _gte checks and used them where it makes sense
* Added version check tests. Again, pretty primitive way of testing.
* Removed unneeded dependencies on travis file

thibaultcha commented on kong-ngx-build:
```diff
@@ -279,7 +279,7 @@ main() {
 
     pushd $OPENSSL_DOWNLOAD
-      if [[ ($OPENSSL_VER == 1.0.* && ! -d include/openssl) || ! -f Makefile ]]; then
+      if (version_lt $OPENSSL_VER 1.1.0 && [[ ! -d include/openssl ]]) || [[ ! -f Makefile ]]; then
```
`version_lte $OPENSSL_VER 1.0` reads better - I can update this during the merge.

thibaultcha commented on t/001_test_version.sh:
```diff
@@ -0,0 +1,82 @@
+t "version_gte 1.11.12 1.11.11" 0
+t "version_gte 1.11.10 1.11.11" 1
+
+exit $g_ok
```
Looks great!

thibaultcha commented on kong-ngx-build:
```diff
@@ -387,6 +393,70 @@ main() {
+  [[ -z $1 || -z $2 ]] && return 1
+
+  # remove non-version parts of a version, ie foobar-1.2.3a.5-rc1 -> 1.2.3a.5
+  local re='^[^0-9]*([0-9a-zA-Z.]*).*$'
```
Maybe just `^[^0-9]*([0-9a-zA-Z.]*)`?

eskerda replied:
But then we do not discard the latter part of the version in `foo-1.2.3-5-rc19`. If we are to effectively parse `-rc19` in `5-rc19` then a couple of the regexes need to change.

thibaultcha replied:
Maybe we are lacking some tests covering these cases?

thibaultcha commented on kong-ngx-build:
```diff
@@ -387,6 +393,70 @@ main() {
 }
 
+parse_version() {
+  [[ -z $1 || -z $2 ]] && return 1
```
We should probably throw a hard error here, as this signifies a programmer error.

thibaultcha commented:
@eskerda It looks like we can merge this very soon, nice job! Do you have a plan for the final commits? I suggest:

1. A commit adding the version utilities (`feat()`)
2. A commit refactoring the `main` function to call the version utilities (this is the `fix()` commit)
3. A commit adding the .travis.yml file (`chore(ci)`)

Or else we can squash them all in a single one, up to you.

thibaultcha commented:
Opened #20 which cleans up the commit history, pending minor comments, and hardens the CI.

eskerda commented:
> Opened #20 which cleans up the commit history, pending minor comments, and hardens the CI.

Looks good! I had the commit history cleaned up, but #20 tests are way way more complete. Let's close this one
